crates/sui-types/src/messages.rs

```diff
@@ -1604,15 +1651,25 @@ impl TransactionData {
         let mut inputs = self.kind.input_objects()?;
 
         if !self.kind.is_system_tx() && !self.kind.is_pay_sui_tx() {
-            inputs.push(InputObjectKind::ImmOrOwnedMoveObject(
-                *self.gas_payment_object_ref(),
-            ));
+            inputs.extend(
+                self.gas()
+                    .iter()
+                    .map(|obj_ref| InputObjectKind::ImmOrOwnedMoveObject(*obj_ref)),
```

Comment:
If you want to be fancy
```suggestion
                    .iter()
                    .copied()
                    .map(InputObjectKind::ImmOrOwnedMoveObject),
```